syntax = "proto3";

package replicacao_dados;

message Data {
  string epoch = 1;
  int32 offset = 2;
  string content = 3;
}

message Ack {
  string message = 1;
}

message CommitOrder {
  string epoch = 1;
  int32 offset = 2;
}

message QueryRequest {
  int32 offset = 1;
}

message QueryResponse {
  string epoch = 1;
  int32 offset = 2;
  string content = 3;
}

message SyncLogRequest {
  string epoch = 1;
  int32 offset = 2;
}

message SyncLogResponse {
  repeated Data entries = 1;
}

service LeaderService {
  rpc ReceiveData(Data) returns (Ack);
  rpc CommitData(CommitOrder) returns (Ack);
  rpc QueryData(QueryRequest) returns (QueryResponse);
  rpc SyncLog(SyncLogRequest) returns (SyncLogResponse);

}

service ReplicaService {
  rpc ReceiveDataFromLeader(Data) returns (Ack);
  rpc CommitData(CommitOrder) returns (Ack);
}